From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Wed, 27 Nov 2019 05:23:02 -0800
Subject: [PATCH] Fix Chunk#getChunkKey(Location)


diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index 7909caef781412b6d7fb1a1add634e63a9bf912b..d21376952b862df3747f74267f77cce089dcdb55 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -40,7 +40,7 @@ public interface Chunk {
      * @return Location's chunk coordinates packed into a long
      */
     static long getChunkKey(@NotNull Location loc) {
-        return getChunkKey((int) Math.floor(loc.getX()) << 4, (int) Math.floor(loc.getZ()) << 4);
+        return getChunkKey((int) Math.floor(loc.getX()) >> 4, (int) Math.floor(loc.getZ()) >> 4); /* Concrete */
     }
 
     /**
-- 
2.24.0

