From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Wed, 19 Jun 2019 13:21:17 -0700
Subject: [PATCH] MC-Dev fixes


diff --git a/src/main/java/net/minecraft/server/RegionFileSection.java b/src/main/java/net/minecraft/server/RegionFileSection.java
index 04b7dab646edd78f344449abef95fb8df05e4860..7eaa720bcf8c4dbc519b8cc971039ee32fb7803f 100644
--- a/src/main/java/net/minecraft/server/RegionFileSection.java
+++ b/src/main/java/net/minecraft/server/RegionFileSection.java
@@ -128,7 +128,7 @@ public class RegionFileSection<R extends MinecraftSerializable> extends RegionFi
             for (int l = 0; l < 16; ++l) {
                 long i1 = SectionPosition.a(chunkcoordintpair, l).v();
                 Optional<R> optional = optionaldynamic.get(Integer.toString(l)).get().map((dynamic2) -> {
-                    return this.e.apply(() -> { // Paper - decompile fix
+                    return this.e.apply(() -> { // Paper - decompile fix /* Concrete - no diff */
                         this.a(i1);
                     }, dynamic2);
                 });
