From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Sat, 18 May 2019 12:25:19 -0700
Subject: [PATCH] Util patch


diff --git a/src/main/java/ca/spottedleaf/concrete/util/Util.java b/src/main/java/ca/spottedleaf/concrete/util/Util.java
new file mode 100644
index 0000000000000000000000000000000000000000..1282e3fb11f996f8626bfb1eaa169da26379c0eb
--- /dev/null
+++ b/src/main/java/ca/spottedleaf/concrete/util/Util.java
@@ -0,0 +1,17 @@
+package ca.spottedleaf.concrete.util;
+
+import org.bukkit.Bukkit;
+
+public final class Util {
+
+    public static void ensureTickThread(final String reason) {
+        if (!Bukkit.isPrimaryThread()) {
+            throw new IllegalStateException(reason);
+        }
+    }
+
+    public static void ensureOwnsChunk(final int x, final int z, final String reason) {
+
+    }
+
+}
