From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Sat, 18 May 2019 12:25:19 -0700
Subject: [PATCH] Util patch


diff --git a/src/main/java/com/destroystokyo/paper/util/map/QueuedChangesMapLong2Object.java b/src/main/java/com/destroystokyo/paper/util/map/QueuedChangesMapLong2Object.java
index 07685b6bd59ba35c8ba53659f6ca8fa1273754ce..7bab31a312463cc963d9621cdc543a281459bd32 100644
--- a/src/main/java/com/destroystokyo/paper/util/map/QueuedChangesMapLong2Object.java
+++ b/src/main/java/com/destroystokyo/paper/util/map/QueuedChangesMapLong2Object.java
@@ -46,10 +46,18 @@ public class QueuedChangesMapLong2Object<V> {
         return this.updatingMap.get(k);
     }
 
+    public boolean updatingContainsKey(final long k) {
+        return this.updatingMap.containsKey(k);
+    }
+
     public V getVisible(final long k) {
         return this.visibleMap.get(k);
     }
 
+    public boolean visibleContainsKey(final long k) {
+        return this.visibleMap.containsKey(k);
+    }
+
     public V getVisibleAsync(final long k) {
         long readlock;
         V ret = null;
@@ -72,6 +80,28 @@ public class QueuedChangesMapLong2Object<V> {
         return ret;
     }
 
+    public boolean visibleContainsKeyAsync(final long k) {
+        long readlock;
+        boolean ret = false;
+
+        do {
+            readlock = this.updatingMapSeqLock.acquireRead();
+
+            try {
+                ret = this.visibleMap.containsKey(k);
+            } catch (final Throwable thr) {
+                if (thr instanceof ThreadDeath) {
+                    throw (ThreadDeath)thr;
+                }
+                // ignore...
+                continue;
+            }
+
+        } while (!this.updatingMapSeqLock.tryReleaseRead(readlock));
+
+        return ret;
+    }
+
     public Long2ObjectLinkedOpenHashMap<V> getVisibleMap() {
         return this.visibleMap;
     }
diff --git a/src/main/java/com/destroystokyo/paper/util/misc/AreaMap.java b/src/main/java/com/destroystokyo/paper/util/misc/AreaMap.java
index f625da9f0968b6f41e302b628c4439198eb4ed64..6fae6d0dbb2256b9c15b69872fe41e7f7a9d525c 100644
--- a/src/main/java/com/destroystokyo/paper/util/misc/AreaMap.java
+++ b/src/main/java/com/destroystokyo/paper/util/misc/AreaMap.java
@@ -77,26 +77,54 @@ public abstract class AreaMap<E> {
         return this.areaMap.size();
     }
 
-    public final void update(final E object, final int chunkX, final int chunkZ, final int viewDistance) {
+    public final void addOrUpdate(final E object, final int chunkX, final int chunkZ, final int viewDistance) {
         final int oldViewDistance = this.objectToViewDistance.put(object, viewDistance);
         final long newPos = MCUtil.getCoordinateKey(chunkX, chunkZ);
+        final long oldPos = this.objectToLastCoordinate.put(object, newPos);
+
         if (oldViewDistance == -1) {
-            this.objectToLastCoordinate.put(object, newPos);
             this.addObject(object, chunkX, chunkZ, Integer.MIN_VALUE, Integer.MIN_VALUE, viewDistance);
             this.addObjectCallback(object, chunkX, chunkZ, viewDistance);
         } else {
+            this.updateObject(object, oldPos, newPos, oldViewDistance, viewDistance);
+            this.updateObjectCallback(object, oldPos, newPos, oldViewDistance, viewDistance);
+        }
+        //this.validate(object, viewDistance);
+    }
+
+    public final boolean update(final E object, final int chunkX, final int chunkZ, final int viewDistance) {
+        final int oldViewDistance = this.objectToViewDistance.replace(object, viewDistance);
+        if (oldViewDistance == -1) {
+            return false;
+        } else {
+            final long newPos = MCUtil.getCoordinateKey(chunkX, chunkZ);
             final long oldPos = this.objectToLastCoordinate.put(object, newPos);
             this.updateObject(object, oldPos, newPos, oldViewDistance, viewDistance);
             this.updateObjectCallback(object, oldPos, newPos, oldViewDistance, viewDistance);
         }
         //this.validate(object, viewDistance);
+        return true;
     }
 
     // called after the distance map updates
-    protected void addObjectCallback(final E object, final int chunkX, final int chunkZ, final int viewDistance) {}
+    protected void updateObjectCallback(final E Object, final long oldPosition, final long newPosition, final int oldViewDistance, final int newViewDistance) {}
+
+    public final boolean add(final E object, final int chunkX, final int chunkZ, final int viewDistance) {
+        final int oldViewDistance = this.objectToViewDistance.putIfAbsent(object, viewDistance);
+        if (oldViewDistance != -1) {
+            return false;
+        }
+
+        final long newPos = MCUtil.getCoordinateKey(chunkX, chunkZ);
+        this.objectToLastCoordinate.put(object, newPos);
+        this.addObject(object, chunkX, chunkZ, Integer.MIN_VALUE, Integer.MIN_VALUE, viewDistance);
+        this.addObjectCallback(object, chunkX, chunkZ, viewDistance);
+
+        return true;
+    }
 
     // called after the distance map updates
-    protected void updateObjectCallback(final E Object, final long oldPosition, final long newPosition, final int oldViewDistance, final int newViewDistance) {}
+    protected void addObjectCallback(final E object, final int chunkX, final int chunkZ, final int viewDistance) {}
 
     public final boolean remove(final E object) {
         final long position = this.objectToLastCoordinate.removeLong(object);
diff --git a/src/main/java/com/destroystokyo/paper/util/misc/pooled/PooledObjects.java b/src/main/java/com/destroystokyo/paper/util/misc/pooled/PooledObjects.java
new file mode 100644
index 0000000000000000000000000000000000000000..25b76af59a334bb8d538f7fbbfdcb16eaf4e0d07
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/util/misc/pooled/PooledObjects.java
@@ -0,0 +1,185 @@
+package com.destroystokyo.paper.util.misc.pooled;
+
+import org.apache.commons.lang3.mutable.MutableInt;
+import java.util.ArrayDeque;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.locks.ReentrantLock;
+
+public final class PooledObjects<E> {
+
+    public static final PooledObjects<MutableInt> POOLED_MUTABLE_INTEGERS = new PooledObjects<>(new PooledObjectHandler<MutableInt>() {
+        @Override
+        public MutableInt createNew() {
+            return new MutableInt();
+        }
+
+        @Override
+        public void onAcquire(final MutableInt value) {}
+
+        @Override
+        public void onRelease(final MutableInt value) {}
+    }, 200, -1);
+
+    private final PooledObjectHandler<E> handler;
+    private final int maxPoolSize;
+    private final int expectingThreads;
+
+    private final IsolatedPool<E> mainPool;
+    // use these under contention
+    private final IsolatedPool<E>[] contendedPools;
+
+    public PooledObjects(final PooledObjectHandler<E> handler, final int maxPoolSize, int expectingThreads) {
+        if (handler == null) {
+            throw new NullPointerException("Handler must not be null");
+        }
+        if (maxPoolSize <= 0) {
+            throw new IllegalArgumentException("Max pool size must be greater-than 0");
+        }
+        if (expectingThreads <= 0) {
+            expectingThreads = Runtime.getRuntime().availableProcessors();
+        }
+
+        this.handler = handler;
+        this.maxPoolSize = maxPoolSize;
+        this.expectingThreads = expectingThreads;
+        this.mainPool = new IsolatedPool<>(handler, maxPoolSize);
+        final IsolatedPool<E>[] contendedPools = new IsolatedPool[2 * expectingThreads];
+
+        for (int i = 0; i < contendedPools.length; ++i) {
+            contendedPools[i] = new IsolatedPool<>(handler, Math.max(1, maxPoolSize / 2));
+        }
+
+        this.contendedPools = contendedPools;
+    }
+
+    // Taken from
+    // https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/
+    // https://github.com/lemire/Code-used-on-Daniel-Lemire-s-blog/blob/master/2016/06/25/fastrange.c
+    // Original license is public domain
+    public static int fastRandomBounded(final long randomInteger, final long limit) {
+        // randomInteger must be [0, pow(2, 32))
+        // limit must be [0, pow(2, 32))
+        return (int)((randomInteger * limit) >>> 32);
+    }
+
+    public E acquire() {
+        E ret;
+        PooledObjects.IsolatedPool<E> pooled = this.mainPool;
+        int lastIndex = -1;
+        while ((ret = pooled.tryAcquireUncontended()) == null) {
+            int index;
+            while (lastIndex == (index = fastRandomBounded(ThreadLocalRandom.current().nextInt() & 0xFFFFFFFFL, this.contendedPools.length)));
+            lastIndex = index;
+            pooled = this.contendedPools[index];
+        }
+
+        return ret;
+    }
+
+    public void release(final E value) {
+        PooledObjects.IsolatedPool<E> pooled = this.mainPool;
+        int lastIndex = -1;
+        while (!pooled.tryReleaseUncontended(value)) {
+            int index;
+            while (lastIndex == (index = fastRandomBounded(ThreadLocalRandom.current().nextInt() & 0xFFFFFFFFL, this.contendedPools.length)));
+            lastIndex = index;
+            pooled = this.contendedPools[index];
+        }
+    }
+
+    /** This object is restricted from interacting with any pool */
+    static interface PooledObjectHandler<E> {
+
+        /**
+         * Must return a non-null object
+         */
+        E createNew();
+
+        void onAcquire(final E value);
+
+        void onRelease(final E value);
+    }
+
+    protected static class IsolatedPool<E> {
+
+        protected final PooledObjectHandler<E> handler;
+
+        // We use arraydeque as it doesn't create garbage per element...
+        protected final ArrayDeque<E> pool;
+        protected final int maxPoolSize;
+
+        protected final ReentrantLock lock = new ReentrantLock();
+
+        public IsolatedPool(final PooledObjectHandler<E> handler, final int maxPoolSize) {
+            this.handler = handler;
+            this.pool = new ArrayDeque<>();
+            this.maxPoolSize = maxPoolSize;
+        }
+
+        protected E acquireOrCreateNoLock() {
+            E ret;
+
+            ret = this.pool.poll();
+
+            if (ret == null) {
+                ret = this.handler.createNew();
+            }
+            this.handler.onAcquire(ret);
+
+            return ret;
+        }
+
+        public E tryAcquireUncontended() {
+            if (!this.lock.tryLock()) {
+                return null;
+            }
+            try {
+                return this.acquireOrCreateNoLock();
+            } finally {
+                this.lock.unlock();
+            }
+        }
+
+        public E acquire() {
+            this.lock.lock();
+            try {
+                return this.acquireOrCreateNoLock();
+            } finally {
+                this.lock.unlock();
+            }
+        }
+
+        protected void releaseNoLock(final E value) {
+            if (this.pool.size() >= this.maxPoolSize) {
+                this.handler.onRelease(value);
+                return; // can't accept, we're at capacity
+            }
+
+            this.pool.add(value);
+            this.handler.onRelease(value);
+        }
+
+        public boolean tryReleaseUncontended(final E value) {
+            if (!this.lock.tryLock()) {
+                return false;
+            }
+
+            try {
+                this.releaseNoLock(value);
+            } finally {
+                this.lock.unlock();
+            }
+
+            return true;
+        }
+
+        public void release(final E value) {
+            this.lock.lock();
+            try {
+                this.releaseNoLock(value);
+            } finally {
+                this.lock.unlock();
+            }
+        }
+    }
+}
diff --git a/src/main/java/net/minecraft/server/BlockPosition.java b/src/main/java/net/minecraft/server/BlockPosition.java
index 2d07d350d2b73197b6ea73cf90ff59679da7e0e7..59d3dc7e810d2497a69ba4180ead9a2e4b7364db 100644
--- a/src/main/java/net/minecraft/server/BlockPosition.java
+++ b/src/main/java/net/minecraft/server/BlockPosition.java
@@ -285,6 +285,7 @@ public class BlockPosition extends BaseBlockPosition implements MinecraftSeriali
             super(i, j, k);
         }
 
+        public static BlockPosition.PooledBlockPosition acquire() { return r(); } // Tuinity - OBFHELPER
         public static BlockPosition.PooledBlockPosition r() {
             return f(0, 0, 0);
         }
diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index 2d4d7250f2ea5db0e522371659541fe03bd6c197..486eb06d9fe511fe67d9ce9e06bfa2cbb910327f 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -750,11 +750,10 @@ public class ChunkProviderServer extends IChunkProvider {
             EnumCreatureType[] aenumcreaturetype = EnumCreatureType.values();
             // Paper start - per player mob spawning
             int[] worldMobCount;
-            if (this.playerChunkMap.playerMobDistanceMap != null) {
+            // Tuinity start - use view distance map
+            if (this.world.paperConfig.perPlayerMobSpawns) {
                 // update distance map
-                this.world.timings.playerMobDistanceMapUpdate.startTiming();
-                this.playerChunkMap.playerMobDistanceMap.update(this.world.players, this.playerChunkMap.viewDistance);
-                this.world.timings.playerMobDistanceMapUpdate.stopTiming();
+                // Tuinity end - use view distance map
                 // re-set mob counts
                 for (EntityPlayer player : this.world.players) {
                     Arrays.fill(player.mobCounts, 0);
@@ -837,9 +836,23 @@ public class ChunkProviderServer extends IChunkProvider {
 
                                     if (this.world.paperConfig.perPlayerMobSpawns) {
                                         int minDiff = Integer.MAX_VALUE;
-                                        for (EntityPlayer entityplayer : this.playerChunkMap.playerMobDistanceMap.getPlayersInRange(chunk.getPos())) {
+                                        // Tuinity start - use view distance map
+                                        com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<EntityPlayer> players = this.playerChunkMap.playerViewDistanceMap.getObjectsInRange(chunk.coordinateKey);
+                                        if (players != null) {
+                                            Object[] backingSet = players.getBackingSet();
+                                            for (int index = 0, len = backingSet.length; index < len; ++index) {
+                                                Object temp = backingSet[index];
+                                                if (!(temp instanceof EntityPlayer)) {
+                                                    continue;
+                                                }
+                                                EntityPlayer entityplayer = (EntityPlayer)temp;
+                                            if (entityplayer.isSpectator() || !entityplayer.affectsSpawning) {
+                                                continue;
+                                            }
+                                            // Tuinity end - use view distance map
                                             minDiff = Math.min(limit - this.playerChunkMap.getMobCountNear(entityplayer, enumcreaturetype), minDiff);
                                         }
+                                        } // Tuinity - use view distance map
                                         difference = (minDiff == Integer.MAX_VALUE) ? 0 : minDiff;
                                     }
 
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 079ddc714e8d431b18e96547fae95582569defbb..eb093558476343838685061e1dbe77583a0ead4c 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -113,6 +113,10 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
 
     public final com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<EntityPlayer> cachedSingleHashSet; // Paper
 
+    // Tuinity start
+    public final com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<EntityPlayer> cachedSingleHashSetTuinity;
+    // Tuinity end
+
     public EntityPlayer(MinecraftServer minecraftserver, WorldServer worldserver, GameProfile gameprofile, PlayerInteractManager playerinteractmanager) {
         super((World) worldserver, gameprofile);
         playerinteractmanager.player = this;
@@ -131,6 +135,9 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         this.canPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
         this.cachedSingleMobDistanceMap = new com.destroystokyo.paper.util.PooledHashSets.PooledObjectLinkedOpenHashSet<>(this); // Paper
+        // Tuinity start
+        this.cachedSingleHashSetTuinity = new com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<>(this);
+        // Tuinity end
     }
 
     // Yes, this doesn't match Vanilla, but it's the best we can do for now.
diff --git a/src/main/java/net/minecraft/server/HeightMap.java b/src/main/java/net/minecraft/server/HeightMap.java
index 8a272ae1b800e31aab6431b647e5b349430b2634..aa73396100adbf741ed52f457284a78c73d658c3 100644
--- a/src/main/java/net/minecraft/server/HeightMap.java
+++ b/src/main/java/net/minecraft/server/HeightMap.java
@@ -119,6 +119,7 @@ public class HeightMap {
         }
     }
 
+    public final int get(int x, int z) { return this.a(x, z); } // Tuinity - OBFHELPER
     public int a(int i, int j) {
         return this.a(c(i, j));
     }
diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
index ee0cabadc8af397522371ab3aaac196f430cfeca..9fed418cdb17335e9cfc19a9a8c964a02ed13964 100644
--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
+++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
@@ -81,7 +81,7 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
     public final Int2ObjectMap<PlayerChunkMap.EntityTracker> trackedEntities;
     private final Queue<Runnable> z;
     int viewDistance; // Paper - private -> package private
-    public final com.destroystokyo.paper.util.PlayerMobDistanceMap playerMobDistanceMap; // Paper
+    //public final com.destroystokyo.paper.util.PlayerMobDistanceMap playerMobDistanceMap; // Paper // Tuinity - replaced by view distance map
 
     // CraftBukkit start - recursion-safe executor for Chunk loadCallback() and unloadCallback()
     public final CallbackExecutor callbackExecutor = new CallbackExecutor();
@@ -111,15 +111,27 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
     // Paper start - distance maps
     private final com.destroystokyo.paper.util.misc.PooledLinkedHashSets<EntityPlayer> pooledLinkedPlayerHashSets = new com.destroystokyo.paper.util.misc.PooledLinkedHashSets<>();
 
+    // Tuinity start - distance maps
+    public final com.destroystokyo.paper.util.misc.PlayerAreaMap playerViewDistanceMap;
+    // Tuinity end - distance maps
+
     void addPlayerToDistanceMaps(EntityPlayer player) {
-        this.updateMaps(player);
+        // Tuinity - we need to add players explicitly now
+        // Tuinity start - distance maps
+        int chunkX = MCUtil.getChunkCoordinate(player.locX());
+        int chunkZ = MCUtil.getChunkCoordinate(player.locZ());
+
+        this.playerViewDistanceMap.add(player, chunkX, chunkZ, this.viewDistance);
+        // Tuinity end - distance maps
 
 
 
     }
 
     void removePlayerFromDistanceMaps(EntityPlayer player) {
-
+        // Tuinity start - distance maps
+        this.playerViewDistanceMap.remove(player);
+        // Tuinity end - distance maps
 
 
 
@@ -128,8 +140,9 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
     void updateMaps(EntityPlayer player) {
         int chunkX = MCUtil.getChunkCoordinate(player.locX());
         int chunkZ = MCUtil.getChunkCoordinate(player.locZ());
-
-
+        // Tuinity start - distance maps
+        this.playerViewDistanceMap.update(player, chunkX, chunkZ, this.viewDistance);
+        // Tuinity end - distance maps
 
 
     }
@@ -201,8 +214,12 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
         this.l = supplier;
         this.m = new VillagePlace(new File(this.w, "poi"), datafixer, this.world); // Paper
         this.setViewDistance(i);
-        this.playerMobDistanceMap = this.world.paperConfig.perPlayerMobSpawns ? new com.destroystokyo.paper.util.PlayerMobDistanceMap() : null; // Paper
+        //this.playerMobDistanceMap = this.world.paperConfig.perPlayerMobSpawns ? new com.destroystokyo.paper.util.PlayerMobDistanceMap() : null; // Paper // Tuinity - distance maps
         this.trackerUpdateDistanceSquared = Math.pow(this.world.paperConfig.trackerUpdateDistance, 2); // Paper
+        // Tuinity start - distance maps
+        com.destroystokyo.paper.util.misc.PooledLinkedHashSets<EntityPlayer> sets = this.pooledLinkedPlayerHashSets;
+        this.playerViewDistanceMap = new com.destroystokyo.paper.util.misc.PlayerAreaMap(sets);
+        // Tuinity end - distance maps
     }
 
     public void updatePlayerMobTypeMap(Entity entity) {
@@ -213,9 +230,23 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
         int chunkZ = (int)Math.floor(entity.locZ()) >> 4;
         int index = entity.getEntityType().getEnumCreatureType().ordinal();
 
-        for (EntityPlayer player : this.playerMobDistanceMap.getPlayersInRange(chunkX, chunkZ)) {
+        // Tuinity start - use view distance map
+        com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<EntityPlayer> players = this.playerViewDistanceMap.getObjectsInRange(chunkX, chunkZ);
+        if (players != null) {
+            Object[] backingSet = players.getBackingSet();
+        for (int i = 0, len = backingSet.length; i < len; ++i) {
+            Object temp = backingSet[i];
+            if (!(temp instanceof EntityPlayer)) {
+                continue;
+            }
+            EntityPlayer player = (EntityPlayer)temp;
+            if (player.isSpectator() || !player.affectsSpawning) {
+                continue;
+            }
+            // Tuinity end - use view distance map
             ++player.mobCounts[index];
         }
+        } // Tuinity - use view distance map
     }
 
     public int getMobCountNear(EntityPlayer entityPlayer, EnumCreatureType enumCreatureType) {
@@ -1411,6 +1442,7 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
             if (!flag1) {
                 this.chunkDistanceManager.a(SectionPosition.a((Entity) entityplayer), entityplayer);
             }
+            this.addPlayerToDistanceMaps(entityplayer); // Tuinity - hook distance map better
         } else {
             SectionPosition sectionposition = entityplayer.K();
 
@@ -1418,6 +1450,7 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
             if (!flag2) {
                 this.chunkDistanceManager.b(sectionposition, entityplayer);
             }
+            this.removePlayerFromDistanceMaps(entityplayer); // Tuinity - hook distance map better
         }
 
         for (int k = i - this.viewDistance; k <= i + this.viewDistance; ++k) {
@@ -1541,7 +1574,6 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
         }
 
         this.updateMaps(entityplayer); // Paper - distance maps
-
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 1a3c7c148bebb304c71a2110fdefdc8c6fae0dfd..939a788392f3d0fec958ca046effb75b090c7244 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1207,9 +1207,11 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
         int k = MathHelper.floor((axisalignedbb.minZ - 2.0D) / 16.0D);
         int l = MathHelper.floor((axisalignedbb.maxZ + 2.0D) / 16.0D);
 
+        ChunkProviderServer chunkProvider = ((ChunkProviderServer)this.chunkProvider); // Tuinity - optimize for loaded chunks
+
         for (int i1 = i; i1 <= j; ++i1) {
             for (int j1 = k; j1 <= l; ++j1) {
-                Chunk chunk = (Chunk)this.getChunkIfLoadedImmediately(i1, j1); // Paper
+                Chunk chunk = chunkProvider.getChunkAtIfLoadedMainThread(i1, j1); // Paper // Tuinity - optimize for loaded chunks
 
                 if (chunk != null) {
                     chunk.a(entity, axisalignedbb, list, predicate);
@@ -1228,9 +1230,11 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
         int l = MathHelper.f((axisalignedbb.maxZ + 2.0D) / 16.0D);
         List<T> list = Lists.newArrayList();
 
+        ChunkProviderServer chunkProvider = ((ChunkProviderServer)this.chunkProvider); // Tuinity - optimize for loaded chunks
+
         for (int i1 = i; i1 < j; ++i1) {
             for (int j1 = k; j1 < l; ++j1) {
-                Chunk chunk = (Chunk)this.getChunkIfLoadedImmediately(i1, j1); // Paper
+                Chunk chunk = chunkProvider.getChunkAtIfLoadedMainThread(i1, j1); // Paper // Tuinity - optimize for loaded chunks
 
                 if (chunk != null) {
                     chunk.a(entitytypes, axisalignedbb, list, predicate);
@@ -1250,10 +1254,11 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
         int l = MathHelper.f((axisalignedbb.maxZ + 2.0D) / 16.0D);
         List<T> list = Lists.newArrayList();
         IChunkProvider ichunkprovider = this.getChunkProvider();
+        ChunkProviderServer chunkProvider = ((ChunkProviderServer)this.chunkProvider); // Tuinity - optimize for loaded chunks
 
         for (int i1 = i; i1 < j; ++i1) {
             for (int j1 = k; j1 < l; ++j1) {
-                Chunk chunk = (Chunk)this.getChunkIfLoadedImmediately(i1, j1); // Paper
+                Chunk chunk = chunkProvider.getChunkAtIfLoadedMainThread(i1, j1); // Paper // Tuinity - optimize for loaded chunks
 
                 if (chunk != null) {
                     chunk.a(oclass, axisalignedbb, list, predicate);
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 4b72641f089fac8bfcab94bc8eba9a06eac11406..778e165c998d20335e2d07fad6b9473f6271cfb3 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1212,7 +1212,7 @@ public class WorldServer extends World {
         }
 
         this.registerEntity(entityplayer);
-        this.getChunkProvider().playerChunkMap.addPlayerToDistanceMaps(entityplayer); // Paper - distance maps
+        // Tuinity - hook distance map better - delay this until PlayerChunkMap#addEntity
     }
 
     // CraftBukkit start
@@ -1423,7 +1423,7 @@ public class WorldServer extends World {
             EntityPlayer entityplayer = (EntityPlayer) entity;
 
             this.players.remove(entityplayer);
-            this.getChunkProvider().playerChunkMap.removePlayerFromDistanceMaps(entityplayer); // Paper - distance maps
+             // Tuinity - hook distance map better - delay until PlayerChunkMap#removeEntity
         }
 
         this.getScoreboard().a(entity);
