From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Sun, 3 Mar 2019 18:25:06 -0800
Subject: [PATCH] Concrete Server Config


diff --git a/src/main/java/ca/spottedleaf/concrete/config/ConcreteConfig.java b/src/main/java/ca/spottedleaf/concrete/config/ConcreteConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..75cb933f9f7070031557bf05ad3a044340a34d9f
--- /dev/null
+++ b/src/main/java/ca/spottedleaf/concrete/config/ConcreteConfig.java
@@ -0,0 +1,153 @@
+package ca.spottedleaf.concrete.config;
+
+import ca.spottedleaf.concurrentutil.util.Throw;
+import org.bukkit.Bukkit;
+import org.bukkit.configuration.ConfigurationSection;
+import org.bukkit.configuration.file.YamlConfiguration;
+
+import java.io.File;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.logging.Level;
+
+public final class ConcreteConfig {
+
+    public static final String CONFIG_HEADER = "Configuration file for Concrete.";
+    public static final int CURRENT_CONFIG_VERSION = 0;
+
+    private static final Object[] EMPTY = new Object[0];
+
+    private static File configFile;
+    private static YamlConfiguration config;
+    private static int configVersion;
+
+    public static void init(final File file) {
+        ConcreteConfig.configFile = file;
+        final YamlConfiguration config = new YamlConfiguration();
+        config.options().header(CONFIG_HEADER);
+        config.options().copyDefaults(true);
+
+        if (!file.exists()) {
+            try {
+                file.createNewFile();
+            } catch (final Exception ex) {
+                Bukkit.getLogger().log(Level.SEVERE, "Failure to create concrete config", ex);
+            }
+        } else {
+            try {
+                config.load(file);
+            } catch (final Exception ex) {
+                Bukkit.getLogger().log(Level.SEVERE, "Failure to load concrete config", ex);
+                Throw.rethrow(ex); /* Rethrow, this is critical */
+                throw new RuntimeException(ex); // unreachable
+            }
+        }
+
+        ConcreteConfig.load(config);
+    }
+
+    public static void load(final YamlConfiguration config) {
+        ConcreteConfig.config = config;
+        ConcreteConfig.configVersion = ConcreteConfig.getInt("config-version-please-do-not-modify-me", CURRENT_CONFIG_VERSION);
+
+        for (final Method method : ConcreteConfig.class.getDeclaredMethods()) {
+            if (method.getReturnType() != void.class || method.getParameterCount() != 0 ||
+                    !Modifier.isPrivate(method.getModifiers()) || !Modifier.isStatic(method.getModifiers())) {
+                continue;
+            }
+
+            try {
+                method.setAccessible(true);
+                method.invoke(null, EMPTY);
+            } catch (final Exception ex) {
+                Throw.rethrow(ex);
+                throw new RuntimeException(ex); // unreachable
+            }
+        }
+
+        /* We re-save to add new options */
+        try {
+            config.save(ConcreteConfig.configFile);
+        } catch (final Exception ex) {
+            Bukkit.getLogger().log(Level.SEVERE, "Unable to save concrete config", ex);
+        }
+    }
+
+    private static boolean getBoolean(final String path, final boolean dfl) {
+        ConcreteConfig.config.addDefault(path, Boolean.valueOf(dfl));
+        return ConcreteConfig.config.getBoolean(path, dfl);
+    }
+
+    private static int getInt(final String path, final int dfl) {
+        ConcreteConfig.config.addDefault(path, Integer.valueOf(dfl));
+        return ConcreteConfig.config.getInt(path, dfl);
+    }
+
+    private static double getDouble(final String path, final double dfl) {
+        ConcreteConfig.config.addDefault(path, Double.valueOf(dfl));
+        return ConcreteConfig.config.getDouble(path, dfl);
+    }
+
+    public static final class WorldConnfig {
+
+        public final String worldName;
+        public ConfigurationSection config;
+
+        public WorldConnfig(final String worldName) {
+            this.worldName = worldName;
+            this.init();
+        }
+
+        public void init() {
+            ConfigurationSection section = ConcreteConfig.config.getConfigurationSection(this.worldName);
+            if (section == null) {
+                section = ConcreteConfig.config.createSection(this.worldName);
+            }
+            ConcreteConfig.config.set(this.worldName, section);
+
+            this.load(section);
+        }
+
+        public void load(final ConfigurationSection config) {
+            this.config = config;
+
+            for (final Method method : ConcreteConfig.WorldConnfig.class.getDeclaredMethods()) {
+                if (method.getReturnType() != void.class || method.getParameterCount() != 0 ||
+                        !Modifier.isPrivate(method.getModifiers()) || Modifier.isStatic(method.getModifiers())) {
+                    continue;
+                }
+
+                try {
+                    method.setAccessible(true);
+                    method.invoke(this, EMPTY);
+                } catch (final Exception ex) {
+                    Throw.rethrow(ex);
+                    throw new RuntimeException(ex); // unreachable
+                }
+            }
+
+            /* We re-save to add new options */
+            try {
+                ConcreteConfig.config.save(ConcreteConfig.configFile);
+            } catch (final Exception ex) {
+                Bukkit.getLogger().log(Level.SEVERE, "Unable to save concrete config", ex);
+            }
+        }
+
+        private boolean getBoolean(final String path, final boolean dfl) {
+            this.config.addDefault(path, Boolean.valueOf(dfl));
+            return this.config.getBoolean(path, dfl);
+        }
+
+        private int getInt(final String path, final int dfl) {
+            this.config.addDefault(path, Integer.valueOf(dfl));
+            return this.config.getInt(path, dfl);
+        }
+
+        private double getDouble(final String path, final double dfl) {
+            this.config.addDefault(path, Double.valueOf(dfl));
+            return this.config.getDouble(path, dfl);
+        }
+    }
+
+}
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index 3ed74ae0ec19a27d96926d14745650df03a8576c..d947faacd076da4cc38b6faa1dd8ac6368af6921 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -186,6 +186,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
         com.destroystokyo.paper.PaperConfig.registerCommands();
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // load version history now
         // Paper end
+        ca.spottedleaf.concrete.config.ConcreteConfig.init((File) options.valueOf("concrete-settings")); /* Concrete - Server Config */
 
         this.setSpawnAnimals(dedicatedserverproperties.spawnAnimals);
         this.setSpawnNPCs(dedicatedserverproperties.spawnNpcs);
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 072dfbffdc5077d445da4c3a115865cc6772672b..d9f357351e591902f2ecf97a9085ebf1478a2f4a 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -96,6 +96,8 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
     public final com.destroystokyo.paper.PaperWorldConfig paperConfig; // Paper
     public final ChunkPacketBlockController chunkPacketBlockController; // Paper - Anti-Xray
 
+    public final ca.spottedleaf.concrete.config.ConcreteConfig.WorldConnfig concreteConfig; /* Concrete - Server Config */
+
     public final co.aikar.timings.WorldTimingsHandler timings; // Paper
     public static BlockPosition lastPhysicsProblem; // Spigot
     private org.spigotmc.TickLimiter entityLimiter;
@@ -145,6 +147,7 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig( worlddata.getName() ); // Spigot
         this.paperConfig = new com.destroystokyo.paper.PaperWorldConfig(worlddata.getName(), this.spigotConfig); // Paper
         this.chunkPacketBlockController = this.paperConfig.antiXray ? new ChunkPacketBlockControllerAntiXray(this.paperConfig) : ChunkPacketBlockController.NO_OPERATION_INSTANCE; // Paper - Anti-Xray
+        this.concreteConfig = new ca.spottedleaf.concrete.config.ConcreteConfig.WorldConnfig(worlddata.getName()); /* Concrete - Server Config */
         this.generator = gen;
         this.world = new CraftWorld((WorldServer) this, gen, env);
         this.ticksPerAnimalSpawns = this.getServer().getTicksPerAnimalSpawns(); // CraftBukkit
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 5f0ce7d5d679504e4ad9222524e422bfc5d8a477..2c6cbd30c385e8cbaed03ddf7969f65c97e7d506 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -794,6 +794,7 @@ public final class CraftServer implements Server {
 
         org.spigotmc.SpigotConfig.init((File) console.options.valueOf("spigot-settings")); // Spigot
         com.destroystokyo.paper.PaperConfig.init((File) console.options.valueOf("paper-settings")); // Paper
+        ca.spottedleaf.concrete.config.ConcreteConfig.init((File) console.options.valueOf("concrete-settings")); /* Concrete - Server Config */
         for (WorldServer world : console.getWorlds()) {
             world.worldData.setDifficulty(config.difficulty);
             world.setSpawnFlags(config.spawnMonsters, config.spawnAnimals);
@@ -810,6 +811,7 @@ public final class CraftServer implements Server {
             }
             world.spigotConfig.init(); // Spigot
             world.paperConfig.init(); // Paper
+            world.concreteConfig.init(); /* Concrete - Server Config */
         }
 
         Plugin[] pluginClone = pluginManager.getPlugins().clone(); // Paper
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 140b82275c7a842d09ecbd2b0c60bd8ab066bc13..2c246780549622ff572a0a2a90bcf00fe2902821 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -136,6 +136,13 @@ public class Main {
                         .defaultsTo(new File("paper.yml"))
                         .describedAs("Yml file");
                 // Paper end
+                /* Concrete Start - Server Config */
+                acceptsAll(asList("concrete", "concrete-settings"), "File for concrete settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("concrete.yml"))
+                        .describedAs("Yml file");
+                /* Conctete End - Server Config */
 
                 // Paper start
                 acceptsAll(asList("server-name"), "Name of the server")
