From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Sat, 5 Jan 2019 11:48:09 -0800
Subject: [PATCH] Use ConcurrentUtil


diff --git a/src/main/java/com/destroystokyo/paper/io/PrioritizedTaskQueue.java b/src/main/java/com/destroystokyo/paper/io/PrioritizedTaskQueue.java
index 78bd238f4ce2a95de5b6bcd455e5e2578e2b8ed8..fffb2d4f616a74c4d52d3989b3c3989fb8756675 100644
--- a/src/main/java/com/destroystokyo/paper/io/PrioritizedTaskQueue.java
+++ b/src/main/java/com/destroystokyo/paper/io/PrioritizedTaskQueue.java
@@ -1,5 +1,6 @@
 package com.destroystokyo.paper.io;
 
+import ca.spottedleaf.concurrentutil.queue.MultiThreadedQueue; /* Concrete - use concurrentutil */
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -48,13 +49,13 @@ public class PrioritizedTaskQueue<T extends PrioritizedTaskQueue.PrioritizedTask
 
     private static final int TOTAL_PRIORITIES     = 6;
 
-    final ConcurrentLinkedQueue<T>[] queues = (ConcurrentLinkedQueue<T>[])new ConcurrentLinkedQueue[TOTAL_PRIORITIES];
+    final MultiThreadedQueue<T>[] queues = new MultiThreadedQueue[TOTAL_PRIORITIES]; /* Concrete - use concurrentutil */
 
     private final AtomicBoolean shutdown = new AtomicBoolean();
 
     {
         for (int i = 0; i < TOTAL_PRIORITIES; ++i) {
-            this.queues[i] = new ConcurrentLinkedQueue<>();
+            this.queues[i] = new MultiThreadedQueue<>(); /* Concrete - use concurrentutil */
         }
     }
 
@@ -73,9 +74,8 @@ public class PrioritizedTaskQueue<T extends PrioritizedTaskQueue.PrioritizedTask
      */
     public void add(final T task) throws IllegalStateException {
         task.onQueue(this);
-        this.queues[task.getPriority()].add(task);
-        if (this.shutdown.get()) {
-            // note: we're not actually sure at this point if our task will go through
+        if (!this.queues[task.getPriority()].add(task)) { /* Concrete - use concurrentutil */
+            // note: we're not actually sure at this point if our task will go through /* Concrete - we are certain now */
             throw new IllegalStateException("Queue has shutdown, refusing to execute task " + IOUtil.genericToString(task));
         }
     }
@@ -86,7 +86,7 @@ public class PrioritizedTaskQueue<T extends PrioritizedTaskQueue.PrioritizedTask
     public T poll() {
         T task;
         for (int i = 0; i < TOTAL_PRIORITIES; ++i) {
-            final ConcurrentLinkedQueue<T> queue = this.queues[i];
+            final MultiThreadedQueue<T> queue = this.queues[i]; /* Concrete - use concurrentutil */
 
             while ((task = queue.poll()) != null) {
                 final int prevPriority = task.tryComplete(i);
@@ -109,7 +109,7 @@ public class PrioritizedTaskQueue<T extends PrioritizedTaskQueue.PrioritizedTask
      */
     public boolean hasTasks() {
         for (int i = 0; i < TOTAL_PRIORITIES; ++i) {
-            final ConcurrentLinkedQueue<T> queue = this.queues[i];
+            final MultiThreadedQueue<T> queue = this.queues[i]; /* Concrete - use concurrentutil */
 
             if (queue.peek() != null) {
                 return true;
@@ -130,6 +130,12 @@ public class PrioritizedTaskQueue<T extends PrioritizedTaskQueue.PrioritizedTask
      * @return {@code true} if the queue was shutdown, {@code false} if it has shut down already
      */
     public boolean shutdown() {
+        /* Concrete start - use concurrentutil */
+        for (int i = 0; i < TOTAL_PRIORITIES; ++i) {
+            final MultiThreadedQueue<T> queue = this.queues[i];
+            queue.preventAdds();
+        }
+        /* Concrete end - use concurrentutil */
         return this.shutdown.getAndSet(false);
     }
 
diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index d947faacd076da4cc38b6faa1dd8ac6368af6921..a68df5aaddbf74c5d47c0be3868262dd53b0a7a1 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -43,7 +43,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
 
     private static final Logger LOGGER = LogManager.getLogger();
     private static final Pattern i = Pattern.compile("^[a-fA-F0-9]{40}$");
-    private final java.util.Queue<ServerCommand> serverCommandQueue = new java.util.concurrent.ConcurrentLinkedQueue<ServerCommand>(); // Paper - use a proper queue
+    private final java.util.Queue<ServerCommand> serverCommandQueue = new ca.spottedleaf.concurrentutil.queue.MultiThreadedQueue<>(); // Paper - use a proper queue /* Concrete - Use a better queue */
     private RemoteStatusListener remoteStatusListener;
     public final RemoteControlCommandListener remoteControlCommandListener;
     private RemoteControlListener remoteControlListener;
diff --git a/src/main/java/net/minecraft/server/IAsyncTaskHandler.java b/src/main/java/net/minecraft/server/IAsyncTaskHandler.java
index f7156acb89373b47b3206b0b131506ea5edd091d..9505be6db5c4c5b2bcc2f4b702d93aa6b131027f 100644
--- a/src/main/java/net/minecraft/server/IAsyncTaskHandler.java
+++ b/src/main/java/net/minecraft/server/IAsyncTaskHandler.java
@@ -13,7 +13,7 @@ public abstract class IAsyncTaskHandler<R extends Runnable> implements Mailbox<R
 
     private final String b;
     private static final Logger LOGGER = LogManager.getLogger();
-    private final Queue<R> d = Queues.newConcurrentLinkedQueue();
+    private final ca.spottedleaf.concurrentutil.queue.MultiThreadedQueue<R> d = new ca.spottedleaf.concurrentutil.queue.MultiThreadedQueue<>(); /* Concrete - Use a better queue */
     private int e;
 
     protected IAsyncTaskHandler(String s) {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 1b5962a76e21234a735b545e1da0283e90831318..7ff317c1e07b785d8674bde50b52f42d5c2a4a99 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -166,7 +166,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
     public org.bukkit.command.RemoteConsoleCommandSender remoteConsole;
     //public ConsoleReader reader; // Paper
     public static int currentTick = 0; // Paper - Further improve tick loop
-    public java.util.Queue<Runnable> processQueue = new java.util.concurrent.ConcurrentLinkedQueue<Runnable>();
+    public final java.util.Queue<Runnable> processQueue = new ca.spottedleaf.concurrentutil.queue.MultiThreadedQueue<>(); /* Concrete - Use CLL */
     public int autosavePeriod;
     public boolean serverAutoSave = false; // Paper
     public File bukkitDataPackFolder;
@@ -1188,9 +1188,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         // CraftBukkit start
         // Run tasks that are waiting on processing
         MinecraftTimings.processQueueTimer.startTiming(); // Spigot
-        while (!processQueue.isEmpty()) {
-            processQueue.remove().run();
-        }
+        ((ca.spottedleaf.concurrentutil.queue.MultiThreadedQueue<Runnable>)processQueue).drain(Runnable::run); /* Concrete - Use efficient drain method */
         MinecraftTimings.processQueueTimer.stopTiming(); // Spigot
 
         MinecraftTimings.timeUpdateTimer.startTiming(); // Spigot // Paper
diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 96a785af27e1924b0cc1959254d4ae3cdd0385a1..0490675ac9f01097f428fde5e4d457bc552adb5e 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -42,7 +42,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
         return new DefaultEventLoopGroup(0, (new ThreadFactoryBuilder()).setNameFormat("Netty Local Client IO #%d").setDaemon(true).build());
     });
     private final EnumProtocolDirection h;
-    private final Queue<NetworkManager.QueuedPacket> packetQueue = Queues.newConcurrentLinkedQueue(); private final Queue<NetworkManager.QueuedPacket> getPacketQueue() { return this.packetQueue; } // Paper - OBFHELPER
+    private final Queue<NetworkManager.QueuedPacket> packetQueue = new ca.spottedleaf.concurrentutil.queue.MultiThreadedQueue<>(); private final Queue<NetworkManager.QueuedPacket> getPacketQueue() { return this.packetQueue; } // Paper - OBFHELPER /* Concrete - Use CLL */
     public Channel channel;
     public SocketAddress socketAddress; public void setSpoofedRemoteAddress(SocketAddress address) { this.socketAddress = address; } // Paper - OBFHELPER
     // Spigot Start
diff --git a/src/main/java/net/minecraft/server/ThreadedMailbox.java b/src/main/java/net/minecraft/server/ThreadedMailbox.java
index aed8277ceb873431a008b45a1b2125553c6fd114..7cf78e7a236b2ffed83a20885f595b55e1d8b037 100644
--- a/src/main/java/net/minecraft/server/ThreadedMailbox.java
+++ b/src/main/java/net/minecraft/server/ThreadedMailbox.java
@@ -17,7 +17,7 @@ public class ThreadedMailbox<T> implements Mailbox<T>, AutoCloseable, Runnable {
     private final String e;
 
     public static ThreadedMailbox<Runnable> a(Executor executor, String s) {
-        return new ThreadedMailbox<>(new PairedQueue.c<>(new ConcurrentLinkedQueue()), executor, s);
+        return new ThreadedMailbox<>(new PairedQueue.c<>(new ca.spottedleaf.concurrentutil.queue.MultiThreadedQueue<>()), executor, s); /* Concrete - use concurrentutil */
     }
 
     public ThreadedMailbox(PairedQueue<? super T, ? extends Runnable> pairedqueue, Executor executor, String s) {
diff --git a/src/main/java/net/minecraft/server/UserCache.java b/src/main/java/net/minecraft/server/UserCache.java
index a072222bc5b8b1b62990bce35034b09ebf270865..18b80f019aee138e9cfdb87382079aaaa97f18d5 100644
--- a/src/main/java/net/minecraft/server/UserCache.java
+++ b/src/main/java/net/minecraft/server/UserCache.java
@@ -43,8 +43,8 @@ public class UserCache {
 
     public static final SimpleDateFormat a = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss Z");
     private static boolean c;
-    private final Map<String, UserCache.UserCacheEntry> d = new java.util.concurrent.ConcurrentHashMap<>();private final Map<String, UserCache.UserCacheEntry> nameCache = d; // Paper - OBFHELPER // Paper
-    private final Map<UUID, UserCache.UserCacheEntry> e = new java.util.concurrent.ConcurrentHashMap<>(); // Paper
+    private final Map<String, UserCache.UserCacheEntry> d = new ca.spottedleaf.concurrentutil.map.SingleWriterMultiReaderHashMap<>();private final Map<String, UserCache.UserCacheEntry> nameCache = d; // Paper - OBFHELPER // Paper /* Concrete - Use concurrentutil */
+    private final Map<UUID, UserCache.UserCacheEntry> e = new ca.spottedleaf.concurrentutil.map.SingleWriterMultiReaderHashMap<>(); // Paper /* Concrete - Use concurrentutil */
     private final Deque<GameProfile> f = new java.util.concurrent.LinkedBlockingDeque<GameProfile>(); // CraftBukkit
     private final GameProfileRepository g;
     protected final Gson b;
-- 
2.24.0.windows.2

