From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Sun, 3 Mar 2019 20:55:07 -0800
Subject: [PATCH] Multi-Threaded ticking CraftBukkit

These are the changes to CB

diff --git a/src/main/java/ca/spottedleaf/tuinity/config/TuinityConfig.java b/src/main/java/ca/spottedleaf/tuinity/config/TuinityConfig.java
index c1b7d0d20531ce933d34f64cd981d07387d7e2f0..da5334fa3ea4631bc331759721e02a32b2d16b1b 100644
--- a/src/main/java/ca/spottedleaf/tuinity/config/TuinityConfig.java
+++ b/src/main/java/ca/spottedleaf/tuinity/config/TuinityConfig.java
@@ -94,6 +94,18 @@ public final class TuinityConfig {
         return TuinityConfig.config.getDouble(path, dfl);
     }
 
+    public static boolean tickWorldsInParallel;
+
+    /**
+     * if tickWorldsInParallel == true, then this value is used as a default only for worlds
+     */
+    public static int tickThreads;
+
+    private static void worldticking() {
+        tickWorldsInParallel = TuinityConfig.getBoolean("tick-worlds-in-parallel", false);
+        tickThreads = TuinityConfig.getInt("server-tick-threads", 1); // will be 4 in the future
+    }
+
     public static final class WorldConnfig {
 
         public final String worldName;
@@ -154,6 +166,14 @@ public final class TuinityConfig {
             this.config.addDefault(path, Double.valueOf(dfl));
             return this.config.getDouble(path, dfl);
         }
+
+        /** ignored if {@link TuinityConfig#tickWorldsInParallel} == false */
+        public int threads;
+
+        private void worldthreading() {
+            final int threads = this.getInt("tick-threads", -1);
+            this.threads = threads == -1 ? TuinityConfig.tickThreads : threads;
+        }
     }
 
 }
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index d6b3b9e7ab780d2e79e2aebf11602cae0868988e..ba0b5f815ea9bdf55a1db5484dbc237a9509c203 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1669,7 +1669,10 @@ public final class CraftServer implements Server {
 
     @Override
     public boolean isPrimaryThread() {
-        return Thread.currentThread().equals(console.serverThread); // Paper - Fix issues with detecting main thread properly
+        // Tuinity start
+        final Thread currThread = Thread.currentThread();
+        return currThread == console.serverThread || currThread instanceof ca.spottedleaf.tuinity.util.TickThread;
+        // Tuinity End
     }
 
     @Override
diff --git a/src/main/java/org/spigotmc/AsyncCatcher.java b/src/main/java/org/spigotmc/AsyncCatcher.java
index 9f7d2ef932ab41cef5d3d0736d20a7c7e4a2c888..10606ed03e88e4e467de2d5424df12466c4e7873 100644
--- a/src/main/java/org/spigotmc/AsyncCatcher.java
+++ b/src/main/java/org/spigotmc/AsyncCatcher.java
@@ -10,7 +10,7 @@ public class AsyncCatcher
 
     public static void catchOp(String reason)
     {
-        if ( enabled && Thread.currentThread() != MinecraftServer.getServer().serverThread )
+        if ( enabled && !org.bukkit.Bukkit.isPrimaryThread() ) // Tuinity
         {
             throw new IllegalStateException( "Asynchronous " + reason + "!" );
         }
