From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Wed, 1 Apr 2020 21:09:51 -0700
Subject: [PATCH] Optimise WorldServer#countMobs

Use loaded entity list instead of entity map

diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 40def90571c8b04e99de891d85f74f6b128b97db..d25b5623feec3a86ee30d0cb3836248ba54b6085 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1153,7 +1153,7 @@ public class WorldServer extends World {
     public int[] countMobs(boolean updatePlayerCounts) {
         int[] ret = new int[EntityPlayer.ENUMCREATURETYPE_TOTAL_ENUMS];
         // Paper end
-        ObjectIterator objectiterator = this.entitiesById.values().iterator();
+        Iterator objectiterator = this.loadedEntities.iterator(); // Tuinity - use loaded entity list
 
         while (objectiterator.hasNext()) {
             Entity entity = (Entity) objectiterator.next();
