From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Sat, 25 Apr 2020 17:10:55 -0700
Subject: [PATCH] Reduce blockpos allocation from pathfinding


diff --git a/src/main/java/net/minecraft/server/PathfinderNormal.java b/src/main/java/net/minecraft/server/PathfinderNormal.java
index d7f0df123ba716ca8587cbfab2c8e94961164909..ec55785af2b432b692d3a3bf4298ffb32489bf3b 100644
--- a/src/main/java/net/minecraft/server/PathfinderNormal.java
+++ b/src/main/java/net/minecraft/server/PathfinderNormal.java
@@ -538,7 +538,7 @@ public class PathfinderNormal extends PathfinderAbstract {
                 if (!iblockdata.a(iblockaccess, blockposition, PathMode.LAND)) {
                     return PathType.BLOCKED;
                 } else {
-                    Fluid fluid = iblockaccess.getFluid(blockposition);
+                    Fluid fluid = iblockdata.getFluid(); // Tuinity - remove another getType call
 
                     return fluid.a((Tag) TagsFluid.WATER) ? PathType.WATER : (fluid.a((Tag) TagsFluid.LAVA) ? PathType.LAVA : PathType.OPEN);
                 }
