From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Tue, 7 Apr 2020 18:55:27 -0700
Subject: [PATCH] Log exceptions for async chunk load failures


diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index 1911ad2fca038ef22931f0d4662f8198566ac7a5..6122a0fb83cc085a828399060a2e2d5077c0aa7a 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -405,7 +405,10 @@ public class ChunkProviderServer extends IChunkProvider {
             // either right -> failure
 
             if (throwable != null) {
-                throw new RuntimeException(throwable);
+                // Tuinity start - thank you once again completablefuture
+                MinecraftServer.LOGGER.fatal("Failed to asynchronously load chunk " + chunkPos, throwable);
+                return;
+                // Tuinity end - thank you once again completablefuture
             }
 
             this.removeTicketAtLevel(TicketType.ASYNC_LOAD, chunkPos, ticketLevel, identifier);
@@ -415,12 +418,19 @@ public class ChunkProviderServer extends IChunkProvider {
 
             if (failure.isPresent()) {
                 // failure
-                throw new IllegalStateException("Chunk failed to load: " + failure.get().toString());
+                // Tuinity start - thank you once again completablefuture
+                MinecraftServer.LOGGER.fatal("Failed to asynchronously load chunk " + chunkPos, new IllegalStateException("Chunk failed to load: " + failure.get().toString()));
+                return;
+                // Tuinity end - thank you once again completablefuture
             }
 
             onComplete.accept(either.left().get());
 
-        }, this.serverThreadQueue);
+        }, this.serverThreadQueue).exceptionally((throwable) -> {                 // Tuinity start - thank you once again completablefuture
+            MinecraftServer.LOGGER.fatal("Failed to asynchronously load chunk " + chunkPos);
+            return null;
+        });
+        // Tuinity end - thank you once again completablefuture
     }
 
     public <T> void addTicketAtLevel(TicketType<T> ticketType, ChunkCoordIntPair chunkPos, int ticketLevel, T identifier) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 2842f592592d49aa9641c3017bba9048bc470aa7..2907b900d5fccf35636dccbbba37331ef7e160c2 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2478,6 +2478,13 @@ public class CraftWorld implements World {
             this.world.doIfNotEntityTicking(() -> ret.complete(chunk == null ? null : chunk.bukkitChunk));
         });
 
+        // Tuinity start - thank you once again completablefuture
+        ret.exceptionally((throwable) -> {
+            net.minecraft.server.MinecraftServer.LOGGER.fatal("Failed to asynchronously load chunk (" + x + "," + z + ")", throwable);
+            return null;
+        });
+        // Tuinity end - thank you once again completablefuture
+
         return ret;
     }
     // Paper end
