From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Wed, 1 Jan 2020 13:29:34 -0800
Subject: [PATCH] Optimise IBlockData#isAir

Assist in inlining the function by caching its result

diff --git a/src/main/java/net/minecraft/server/Block.java b/src/main/java/net/minecraft/server/Block.java
index e29ec958b3519d92cda215a50e97e6852d71c684..44e3e1bb4921e6cf4b40fbc5279d05964a8a520c 100644
--- a/src/main/java/net/minecraft/server/Block.java
+++ b/src/main/java/net/minecraft/server/Block.java
@@ -109,8 +109,8 @@ public class Block implements IMaterial {
         return iblockdata.d(iblockaccess, blockposition, EnumDirection.UP) && this.n < 14;
     }
 
-    @Deprecated
-    public boolean d(IBlockData iblockdata) {
+    @Deprecated public final boolean isAir(IBlockData iblockdata) { return this.d(iblockdata); } // Tuinity - OBFHELPER
+    @Deprecated public boolean d(IBlockData iblockdata) { // Tuinity - OBFHELPER
         return false;
     }
 
diff --git a/src/main/java/net/minecraft/server/IBlockData.java b/src/main/java/net/minecraft/server/IBlockData.java
index 05f5cd4a42250a4b988b8b4ece3d86efe3b257a7..60cc055b5b749a2d29470d831561221e191db964 100644
--- a/src/main/java/net/minecraft/server/IBlockData.java
+++ b/src/main/java/net/minecraft/server/IBlockData.java
@@ -22,11 +22,13 @@ public class IBlockData extends BlockDataAbstract<Block, IBlockData> implements
     private IBlockData.a c;
     private final int d;
     private final boolean e;
+    private final boolean isAir;
 
     public IBlockData(Block block, ImmutableMap<IBlockState<?>, Comparable<?>> immutablemap) {
         super(block, immutablemap);
         this.d = block.a(this);
         this.e = block.o(this);
+        this.isAir = this.getBlock().isAir(this); // Tuinity
         this.isTicking = this.getBlock().isTicking(this); // Tuinity
     }
 
@@ -83,7 +85,7 @@ public class IBlockData extends BlockDataAbstract<Block, IBlockData> implements
     }
 
     public boolean isAir() {
-        return this.getBlock().d(this);
+        return this.isAir; // Tuinity - optimise this function
     }
 
     public MaterialMapColor c(IBlockAccess iblockaccess, BlockPosition blockposition) {
