From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Wed, 27 Nov 2019 04:54:58 -0800
Subject: [PATCH] Improve inline of IBlockData#q

(q = isTicking)
Hot method for random chunk ticking

diff --git a/src/main/java/net/minecraft/server/IBlockData.java b/src/main/java/net/minecraft/server/IBlockData.java
index 21f734a73c485b37315d59e8a4d5bcf27989fea4..51ae60f8d1f107b85d822646e175485efeb9c148 100644
--- a/src/main/java/net/minecraft/server/IBlockData.java
+++ b/src/main/java/net/minecraft/server/IBlockData.java
@@ -27,6 +27,7 @@ public class IBlockData extends BlockDataAbstract<Block, IBlockData> implements
         super(block, immutablemap);
         this.d = block.a(this);
         this.e = block.n(this);
+        this.isTicking = this.getBlock().isTicking(this); /* Concrete */
     }
 
     public void c() {
@@ -267,8 +268,11 @@ public class IBlockData extends BlockDataAbstract<Block, IBlockData> implements
         return this.getBlock().g(this);
     }
 
+    /* Concrete start */
+    private final boolean isTicking;
     public boolean q() {
-        return this.getBlock().isTicking(this);
+        return this.isTicking;
+        /* Concrete end */
     }
 
     public final SoundEffectType getStepSound() { return this.r(); } // Paper - OBFHELPER
-- 
2.24.0

