From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Wed, 1 Apr 2020 21:09:51 -0700
Subject: [PATCH] Optimise WorldServer#countMobs

Use loaded entity list instead of entity map

diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 6e04905915b80dc08f44dbef7ec8006a40711fff..6470842907d4242603cd8645838633f12a988d7b 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1470,7 +1470,7 @@ public class WorldServer extends World {
     public int[] countMobs(boolean updatePlayerCounts) {
         int[] ret = new int[EntityPlayer.ENUMCREATURETYPE_TOTAL_ENUMS];
         // Paper end
-        ObjectIterator objectiterator = this.entitiesById.values().iterator();
+        Iterator objectiterator = this.loadedEntities.iterator(); // Tuinity - use loaded entity list
 
         while (objectiterator.hasNext()) {
             Entity entity = (Entity) objectiterator.next();
