From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Wed, 1 Apr 2020 21:09:51 -0700
Subject: [PATCH] Optimise WorldServer#countMobs

Use loaded entity list instead of entity map

diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index adb67cbda7bc39c73d15c63e0595ea44a19b1a48..471b369797276327f217e9acd5bc12e8e4d461e1 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1553,7 +1553,7 @@ public class WorldServer extends World {
     public int[] countMobs(boolean updatePlayerCounts) {
         int[] ret = new int[EntityPlayer.ENUMCREATURETYPE_TOTAL_ENUMS];
         // Paper end
-        ObjectIterator objectiterator = this.entitiesById.values().iterator();
+        Iterator objectiterator = this.loadedEntities.iterator(); // Tuinity - use loaded entity list
 
         while (objectiterator.hasNext()) {
             Entity entity = (Entity) objectiterator.next();
