From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Wed, 1 Apr 2020 21:09:51 -0700
Subject: [PATCH] Optimise WorldServer#countMobs

Use loaded entity list instead of entity map

diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 90b9e7febd87db7802e6c4ce9465fd1a68dfe705..02807c1bb874dc3d27b73ad8b5df34506ae3f3d7 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1552,7 +1552,7 @@ public class WorldServer extends World {
     public int[] countMobs(boolean updatePlayerCounts) {
         int[] ret = new int[EntityPlayer.ENUMCREATURETYPE_TOTAL_ENUMS];
         // Paper end
-        ObjectIterator objectiterator = this.entitiesById.values().iterator();
+        Iterator objectiterator = this.loadedEntities.iterator(); // Tuinity - use loaded entity list
 
         while (objectiterator.hasNext()) {
             Entity entity = (Entity) objectiterator.next();
