From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Tue, 19 May 2020 21:19:15 -0700
Subject: [PATCH] Prevent entities from moving into unloaded chunks

Temporary change until dev/voxel-optimise2 is merged.

diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 352e8699f16f04153f3065b6fd7f3bdadd96aeed..de7e553fdc04dc497e81295d5ed86b24bbebbd22 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -890,6 +890,11 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
     }
 
     private Vec3D e(Vec3D vec3d) {
+        // Tuinity start - prevent entities from moving into unloaded chunks
+        if (!((WorldServer)this.world).areChunksLoadedForMove(this.getBoundingBox().expand(vec3d.x, vec3d.y, vec3d.z))) {
+            return new Vec3D(0.0, 0.0, 0.0);
+        }
+        // Tuinity end - prevent entities from moving into unloaded chunks
         AxisAlignedBB axisalignedbb = this.getBoundingBox();
         VoxelShapeCollision voxelshapecollision = VoxelShapeCollision.a(this);
         VoxelShape voxelshape = this.world.getWorldBorder().a();
