From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Tue, 9 Jun 2020 00:45:01 -0700
Subject: [PATCH] Prevent getChunkAtAsync calls in an unload event from firing
 a load event


diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index 397cdd5451335914705afc4f03c8fed47b20b82e..7aef5bb6b0d6c533bd7f8c795e39627ecfc93356 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -700,7 +700,7 @@ public class ChunkProviderServer extends IChunkProvider {
             PlayerChunk.State currentChunkState = PlayerChunk.getChunkState(playerchunk.getTicketLevel());
             currentlyUnloading = (oldChunkState.isAtLeast(PlayerChunk.State.BORDER) && !currentChunkState.isAtLeast(PlayerChunk.State.BORDER));
         }
-        if (flag) { // Paper - don't care about unloading state
+        if (flag && !currentlyUnloading) { // Paper - don't care about unloading state // Tuinity - prevent firing chunk load events while in chunk unload events
             // CraftBukkit end
             this.chunkMapDistance.a(TicketType.UNKNOWN, chunkcoordintpair, l, chunkcoordintpair);
             if (isUrgent) this.chunkMapDistance.markUrgent(chunkcoordintpair); // Paper
