From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Mon, 6 Jul 2020 22:48:48 -0700
Subject: [PATCH] Manually inline methods in BlockPosition


diff --git a/src/main/java/net/minecraft/server/BaseBlockPosition.java b/src/main/java/net/minecraft/server/BaseBlockPosition.java
index 1842e69839a65e7b8d2c47fd30ec8f3179f04470..dab89b0c69ddff24d37cee06107bdc464012c6bb 100644
--- a/src/main/java/net/minecraft/server/BaseBlockPosition.java
+++ b/src/main/java/net/minecraft/server/BaseBlockPosition.java
@@ -16,9 +16,9 @@ public class BaseBlockPosition implements Comparable<BaseBlockPosition> {
         return IntStream.of(new int[]{baseblockposition.getX(), baseblockposition.getY(), baseblockposition.getZ()});
     });
     public static final BaseBlockPosition ZERO = new BaseBlockPosition(0, 0, 0);
-    private int a;
-    private int b;
-    private int e;
+    protected int a; // Tuinity - private->protected - diff on change, this is the x coordinate
+    protected int b; // Tuinity - private->protected - diff on change, this is the y coordinate
+    protected int e; // Tuinity - private->protected - diff on change, this is the z coordinate
 
     // Paper start
     public boolean isValidLocation() {
@@ -71,15 +71,15 @@ public class BaseBlockPosition implements Comparable<BaseBlockPosition> {
         return this.e;
     }
 
-    protected void o(int i) {
+    protected void o_unused(int i) { // Tuinity - not needed here
         this.a = i;
     }
 
-    protected void p(int i) {
+    protected void p_unused(int i) { // Tuinity - not needed here
         this.b = i;
     }
 
-    protected void q(int i) {
+    protected void q_unused(int i) { // Tuinity - not needed here
         this.e = i;
     }
 
diff --git a/src/main/java/net/minecraft/server/BlockPosition.java b/src/main/java/net/minecraft/server/BlockPosition.java
index 8c2a4b57aba17b43517f7b09c4adf65ae5904b6d..7ff4948a47be03c65c162ca04b2338cdcdafe7c4 100644
--- a/src/main/java/net/minecraft/server/BlockPosition.java
+++ b/src/main/java/net/minecraft/server/BlockPosition.java
@@ -386,11 +386,11 @@ public class BlockPosition extends BaseBlockPosition {
             return super.a(enumblockrotation).immutableCopy();
         }
 
-        public BlockPosition.MutableBlockPosition setValues(int i, int j, int k) { return d(i, j, k);} // Paper - OBFHELPER
-        public BlockPosition.MutableBlockPosition d(int i, int j, int k) {
-            this.o(i);
-            this.p(j);
-            this.q(k);
+        public final BlockPosition.MutableBlockPosition setValues(int i, int j, int k) { return d(i, j, k);} // Paper - OBFHELPER // Tuinity
+        public final BlockPosition.MutableBlockPosition d(int i, int j, int k) { // Tuinity
+            ((BaseBlockPosition)this).a = i; // Tuinity - force inline
+            ((BaseBlockPosition)this).b = j; // Tuinity - force inline
+            ((BaseBlockPosition)this).e = k; // Tuinity - force inline
             return this;
         }
 
@@ -400,12 +400,18 @@ public class BlockPosition extends BaseBlockPosition {
         }
 
         public final BlockPosition.MutableBlockPosition setValues(final BaseBlockPosition baseblockposition) { return this.g(baseblockposition); } // Paper - OBFHELPER
-        public BlockPosition.MutableBlockPosition g(BaseBlockPosition baseblockposition) {
-            return this.d(baseblockposition.getX(), baseblockposition.getY(), baseblockposition.getZ());
+        public final BlockPosition.MutableBlockPosition g(BaseBlockPosition baseblockposition) { // Tuinity
+            ((BaseBlockPosition)this).a = baseblockposition.a; // Tuinity - force inline
+            ((BaseBlockPosition)this).b = baseblockposition.b; // Tuinity - force inline
+            ((BaseBlockPosition)this).e = baseblockposition.e; // Tuinity - force inline
+            return this;
         }
 
-        public BlockPosition.MutableBlockPosition g(long i) {
-            return this.d(b(i), c(i), d(i));
+        public final BlockPosition.MutableBlockPosition g(long i) { // Tuinity
+            ((BaseBlockPosition)this).a = (int)(i >> 38); // Tuinity - force inline
+            ((BaseBlockPosition)this).b = (int)((i << 52) >> 52); // Tuinity - force inline
+            ((BaseBlockPosition)this).e = (int)((i << 26) >> 38); // Tuinity - force inline
+            return this;
         }
 
         public BlockPosition.MutableBlockPosition a(EnumAxisCycle enumaxiscycle, int i, int j, int k) {
@@ -420,8 +426,11 @@ public class BlockPosition extends BaseBlockPosition {
             return this.d(baseblockposition.getX() + i, baseblockposition.getY() + j, baseblockposition.getZ() + k);
         }
 
-        public BlockPosition.MutableBlockPosition c(EnumDirection enumdirection) {
-            return this.c(enumdirection, 1);
+        public final BlockPosition.MutableBlockPosition c(EnumDirection enumdirection) { // Tuinity
+            ((BaseBlockPosition)this).a += enumdirection.getAdjacentX(); // Tuinity - force inline
+            ((BaseBlockPosition)this).b += enumdirection.getAdjacentY(); // Tuinity - force inline
+            ((BaseBlockPosition)this).e += enumdirection.getAdjacentZ(); // Tuinity - force inline
+            return this;
         }
 
         public BlockPosition.MutableBlockPosition c(EnumDirection enumdirection, int i) {
@@ -445,22 +454,31 @@ public class BlockPosition extends BaseBlockPosition {
             }
         }
 
-        public final void setX(final int x) { super.o(x); } // Paper - OBFHELPER
-        @Override
-        public void o(int i) {
-            super.o(i);
+        // Tuinity start
+        public final void setX(int value) {
+            ((BaseBlockPosition)this).a = value;
+        }
+        public final void setY(int value) {
+            ((BaseBlockPosition)this).b = value;
         }
+        public final void setZ(int value) {
+            ((BaseBlockPosition)this).e = value;
+        }
+        // Tuinity end
 
-        public final void setY(final int y) { super.p(y); } // Paper - OBFHELPER
-        @Override
-        public void p(int i) {
-            super.p(i);
+        // Tuinity - moved up
+        public final void o(int i) { // Tuinity
+            ((BaseBlockPosition)this).a = i; // need cast thanks to name conflict
         }
 
-        public final void setZ(final int z) { super.q(z); } // Paper - OBFHELPER
-        @Override
-        public void q(int i) {
-            super.q(i);
+        // Tuinity - moved up
+        public final void p(int i) { // Tuinity
+            ((BaseBlockPosition)this).b = i; // Tuinity
+        }
+
+        // Tuinity - moved up
+        public final void q(int i) { // Tuinity
+            ((BaseBlockPosition)this).e = i; // Tuinity
         }
 
         @Override
