From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Tue, 11 Aug 2020 04:52:24 -0700
Subject: [PATCH] Use visible chunks for light engine status checks

Not mt-safe to use getUpdating

diff --git a/src/main/java/net/minecraft/server/LightEngineThreaded.java b/src/main/java/net/minecraft/server/LightEngineThreaded.java
index ec8498c3db7a8d417f27eed5d3af38cb726b95b6..ceaa92400678aef3bf436279dd807e4b38609dda 100644
--- a/src/main/java/net/minecraft/server/LightEngineThreaded.java
+++ b/src/main/java/net/minecraft/server/LightEngineThreaded.java
@@ -20,7 +20,7 @@ public class LightEngineThreaded extends LightEngine implements AutoCloseable {
     private static final int MAX_PRIORITIES = PlayerChunkMap.GOLDEN_TICKET + 2;
 
     private boolean isChunkLightStatus(long pair) {
-        PlayerChunk playerChunk = playerChunkMap.getUpdatingChunk(pair);
+        PlayerChunk playerChunk = playerChunkMap.getVisibleChunk(pair); // Tuinity - mt-safe access
         if (playerChunk == null) {
             return false;
         }
