From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Tue, 11 Aug 2020 04:52:24 -0700
Subject: [PATCH] Use visible chunks for light engine status checks

Not mt-safe to use getUpdating

diff --git a/src/main/java/net/minecraft/server/LightEngineThreaded.java b/src/main/java/net/minecraft/server/LightEngineThreaded.java
index eab244473e4f82bd8ca1ad9c467f4ff6648b4fed..2cacb948b3e0f7c2500c0fc7cdb07ae473004814 100644
--- a/src/main/java/net/minecraft/server/LightEngineThreaded.java
+++ b/src/main/java/net/minecraft/server/LightEngineThreaded.java
@@ -19,7 +19,7 @@ public class LightEngineThreaded extends LightEngine implements AutoCloseable {
     private static final int MAX_PRIORITIES = PlayerChunkMap.GOLDEN_TICKET + 2;
 
     private boolean isChunkLightStatus(long pair) {
-        PlayerChunk playerChunk = playerChunkMap.getUpdatingChunk(pair);
+        PlayerChunk playerChunk = playerChunkMap.getVisibleChunk(pair); // Tuinity - mt-safe access
         if (playerChunk == null) {
             return false;
         }
