From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Sun, 30 Aug 2020 10:58:55 -0700
Subject: [PATCH] Do not run vanilla update logic when eigencraft is enabled


diff --git a/src/main/java/net/minecraft/server/BlockRedstoneWire.java b/src/main/java/net/minecraft/server/BlockRedstoneWire.java
index 6abc3d4cfbe8f5cc1dbe83031d90ee8df3b2f914..a8ef41dcf55b2f4b45d3000f9b973f7377182750 100644
--- a/src/main/java/net/minecraft/server/BlockRedstoneWire.java
+++ b/src/main/java/net/minecraft/server/BlockRedstoneWire.java
@@ -226,6 +226,7 @@ public class BlockRedstoneWire extends Block {
     private void updateSurroundingRedstone(World worldIn, BlockPosition pos, IBlockData state, BlockPosition source) {
         if (worldIn.paperConfig.useEigencraftRedstone) {
             turbo.updateSurroundingRedstone(worldIn, pos, state, source);
+            return; // Tuinity - Don't run old logic as well
         }
         a(worldIn, pos, state);
     }
