From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Thu, 24 Jun 2021 01:19:04 -0700
Subject: [PATCH] Do not cancel chunk load on ticket level change

Now the chunk system cannot handle cancellations. Yee haw.

diff --git a/src/main/java/net/minecraft/server/level/ChunkHolder.java b/src/main/java/net/minecraft/server/level/ChunkHolder.java
index 6e7df2288cdcf409073e200d6c6efe28ca2f1868..7c5848a421be02d17ce273b2158f0ed1b8bfc8fd 100644
--- a/src/main/java/net/minecraft/server/level/ChunkHolder.java
+++ b/src/main/java/net/minecraft/server/level/ChunkHolder.java
@@ -700,7 +700,7 @@ public class ChunkHolder {
 
             // Paper start
             if (!flag1) {
-                chunkStorage.level.asyncChunkTaskManager.cancelChunkLoad(this.pos.x, this.pos.z);
+                //chunkStorage.level.asyncChunkTaskManager.cancelChunkLoad(this.pos.x, this.pos.z); // Tuinity - cancelling this is a bad idea now
             }
             // Paper end
 
