From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Wed, 17 Mar 2021 03:23:18 -0700
Subject: [PATCH] fixup! Rewrite the light engine


diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
index 3e0bf6df7c4608a5b19024612db52558fd722f4b..945d16f57317406bff3cbe20756677eccfba1525 100644
--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
+++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
@@ -1206,7 +1206,7 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
                     if (ichunkaccess.getChunkStatus().b(chunkstatus)) {
                         CompletableFuture<Either<IChunkAccess, PlayerChunk.Failure>> completablefuture1; // Paper
 
-                        if (chunkstatus == ChunkStatus.LIGHT) {
+                        if (!com.tuinity.tuinity.config.TuinityConfig.useNewLightEngine && chunkstatus == ChunkStatus.LIGHT) { // Tuinity - we use edge checks, so loading 1 radius neighbours isn't necessary
                             completablefuture1 = this.b(playerchunk, chunkstatus);
                         } else {
                             completablefuture1 = chunkstatus.a(this.world, this.definedStructureManager, this.lightEngine, (ichunkaccess1) -> {
