From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Fri, 26 Mar 2021 14:10:58 -0700
Subject: [PATCH] Make sure to remove correct TE during TE tick

This looks like it can cause premature TE removal

diff --git a/src/main/java/net/minecraft/world/level/World.java b/src/main/java/net/minecraft/world/level/World.java
index c8a5d4972431ce9615312280f36181a2b9645df7..68fa071fc576f398682ef461df102be432cdcb4c 100644
--- a/src/main/java/net/minecraft/world/level/World.java
+++ b/src/main/java/net/minecraft/world/level/World.java
@@ -968,7 +968,7 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
                 //this.tileEntityList.remove(tileentity); // Paper - remove unused list
                 // Paper - prevent double chunk lookups
                 Chunk chunk; if ((chunk = this.getChunkIfLoaded(tileentity.getPosition())) != null) { // inlined contents of this.isLoaded(BlockPosition). Reuse the returned chunk instead of looking it up again
-                    chunk.removeTileEntity(tileentity.getPosition());
+                    chunk.removeTileEntity(tileentity.getPosition(), tileentity); // Tuinity - make sure we remove the correct TE
                 }
                 // Paper end
             }
diff --git a/src/main/java/net/minecraft/world/level/chunk/Chunk.java b/src/main/java/net/minecraft/world/level/chunk/Chunk.java
index 259d4ac89e84fd334ff65ea8a606e1fc50cc882b..ffef28f9fa82a6961ef6db5f6732cfee4352ee01 100644
--- a/src/main/java/net/minecraft/world/level/chunk/Chunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/Chunk.java
@@ -956,10 +956,18 @@ public class Chunk implements IChunkAccess {
 
     @Override
     public void removeTileEntity(BlockPosition blockposition) {
+        // Tuinity start - make sure we remove the correct TE
+        this.removeTileEntity(blockposition, null);
+    }
+    public void removeTileEntity(BlockPosition blockposition, TileEntity match) {
+        // Tuinity end - make sure we remove the correct TE
         if (this.loaded || this.world.s_()) {
-            TileEntity tileentity = (TileEntity) this.tileEntities.remove(blockposition);
+            // Tuinity start - make sure we remove the correct TE
+            TileEntity tileentity = (TileEntity) this.tileEntities.get(blockposition);
 
-            if (tileentity != null) {
+            if (tileentity != null && (match == null || match == tileentity)) {
+                this.tileEntities.remove(blockposition);
+                // Tuinity end - make sure we remove the correct TE
                 tileentity.al_();
             }
         }
